visions/abi-stability.md

```diff
@@ -0,0 +1,248 @@
+non-fragile base class support, the Darwin linker has a number of
+features intended to allow code to be moved from one framework or
+dynamic library to another without an ABI break, including support for
+re-exporting symbols, as well as a variety of meta symbols that allow
+for manipulation of symbol visibility and indeed apparent location
+based on system versions.
+
+### Windows
+
+Like Apple platforms, Windows is ABI stable at the API layer; SPIs are
+not officially ABI stable (though in practice they may be), and
+neither is the system call interface itself.
+
+Windows has a platform-wide exception handling mechanism (Structured
+Exception Handling or SEH), which means that the operating system
+specifies in detail how stack unwinding must take place.  Unlike other
+platforms we care about, Windows does not use DWARF for exception
+unwinding, but instead relies on knowledge of the standard function
+prologue and epilogue.  This creates problems for tail call
+optimization, which is also needed for Swift Concurrency.  While we
+have *a* solution in place, we need to be convinced that it's the
+*right* solution before declaring ABI stability on Windows.
+
+Windows DLLs also have some interesting design features that need to
+be borne in mind here.  In particular, symbols exported from DLLs can
+be exported by 16-bit _ordinal_ as well as by name; name exports are
+actually done by mapping the name to an ordinal, and then looking up
+the ordinal in the export table.  If symbols are exported by ordinal
+alone, the ordinals need to be stable in order to achieve ABI
+stability.  Further, because of the way linking works, people have in
+the past generated import libraries directly from DLL files, which
+tends to result in import-by-ordinal rather than import-by-name, which
+has actually forced Microsoft to stabilise ordinals when it otherwise
+would not have done so.
+
+Windows DLLs are also, like dylibs on Darwin, able to re-export
+symbols from other DLLs (Windows calls this feature _forwarding_); the
+Windows feature is also able to rename the symbol at the same time,
+and indeed the Win32 layer re-exports some NT SPIs directly as Win32
+APIs using this mechanism.
+
+As regards data structure ABI stability, Windows uses two different
+mechanisms here; the first is that data structures typically start
+with a header that contains a size field (which gets used,
+essentially, as a version).  The second is use of the Component Object
+Model (COM), which relies on the way the Microsoft compiler lays out
+C++ vtables.
+
+Windows doesn't have anything quite like the `availability` mechanism
+Apple platforms use, though there are C preprocessor macros like
+`WINVER` or `_WIN32_WINNT` that control exactly which versions of data
+structures and which functions are exported by the Windows SDK headers.
+
+Windows programs that need to cope with the possibility that a function
+might not be present at runtime need to use `GetProcAddress()` (the
```

Comment:
```suggestion
might not be present at runtime need to use `GetProcAddress` (the
```

The parenthesis makes it feel like there are no arguments to the invocation.

Reply:
The parentheses just indicate that we're talking about a function. I don't think it's an unusual style, marking function names with parentheses like that (pretty sure I've seen documentation written like that all over the place).